### PR TITLE
feat: Update pipeline to use `trip_id` as unique key

### PR DESCRIPTION
**Pull request by rymarczy**
This change modifies the data processing pipeline to use `trip_id` in the unique identifier for trip-stops.

This is required to process GTFS-RT Vehicle Postion files that do no report `start_date` and `start_time` values. 

The `pipeline_flat_out.csv` test file required 2 updates because of this pipeline change:

- Trip ID `55713710` vehicle label field switch from `3868-3696` to `3696-3819-3662` (The trip reports both labels evenly for the duration of the trip` 
- ~~Trip ID `1683547929` parent station `place-pktrm` trunk headway changed to 410 seconds to match branch headway duration~~

Asana Task: https://app.asana.com/0/1204931901750665/1205259099679735

### CODE REVIEW

File: python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py [hunk 1]
```diff
@@ -371,9 +371,7 @@ def upgrade() -> None:
         sa.UniqueConstraint(
             "service_date",
             "route_id",
-            "direction_id",
-            "start_time",
-            "vehicle_id",
+            "trip_id",
```
Comment (mzappitello) [inline python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py:374]: do we need to have `route_id` on here?

File: python_src/src/lamp_py/performance_manager/l1_rt_trips.py [hunk 2]
```diff
@@ -93,15 +93,13 @@ def load_new_trip_data(db_manager: DatabaseManager) -> None:
...
-            TempEventCompare.vehicle_id,
+            TempEventCompare.trip_id,
         )
         .order_by(
             TempEventCompare.service_date,
             TempEventCompare.route_id,
-            TempEventCompare.direction_id,
-            TempEventCompare.start_time,
+            TempEventCompare.trip_id,
+            TempEventCompare.stop_sequence.desc(),
```
Comment (paulswartz) [inline python_src/src/lamp_py/performance_manager/l1_rt_trips.py:102]: question: is this backwards? stop sequences start with low numbers and go up.

**Review by mzappitello — COMMENTED** [review]
a few comments in here, some questions some suggestions.

seems like our csv stuff isn't lining up tho.

File: python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rymarczy) [reply python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py:374]: We have examples in our test set where the same `trip_id` continues across more than one Green Line `route_id` for the same `start_date`. 

I'm not sure how common this is, but it does occur, and I believe the unique trip designations we are looking for should include `route_id` as a guarantee.

File: python_src/src/lamp_py/performance_manager/l1_rt_trips.py (hunk 2 above)
Comment (rymarczy) [reply python_src/src/lamp_py/performance_manager/l1_rt_trips.py:102]: The purpose of this query is to collect additional information fields ( `vehicle_label`, `vehicle_consist`) for unique trips that will be used for UPDATE/INSERT operations into our `vehicle_trips` table. 

For this first stop-event in a trip, this information is frequently carried over from the last trip for the vehicle, so this query is collecting the latest stop-event values from these information fields to UPDATE/INSERT into our `vehicle_trips` table. 

That is why the `desc()` ORDER is used, it would probably be helpful to add a comment to this effect above these calls.

File: python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rymarczy) [reply python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py:374]: We have examples in our test set where the same trip_id continues across more than one Green Line route_id for the same start_date.

I'm not sure how common this is, but it does occur, and I believe the unique trip designations we are looking for should include route_id as a guarantee.

File: python_src/src/lamp_py/performance_manager/l1_rt_trips.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rymarczy) [reply python_src/src/lamp_py/performance_manager/l1_rt_trips.py:102]: The purpose of this query is to collect additional information fields ( vehicle_label, vehicle_consist) for unique trips that will be used for UPDATE/INSERT operations into our vehicle_trips table.

For this first stop-event in a trip, this information is frequently carried over from the last trip for the vehicle, so this query is collecting the latest stop-event values from these information fields to UPDATE/INSERT into our vehicle_trips table.

That is why the desc() ORDER is used, it would probably be helpful to add a comment to this effect above these calls.

File: python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rymarczy) [reply python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py:374]: Comment has been added to the function call to describe this behavior

[the remark this answers is not in the document]
Comment (paulswartz) [reply python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py:374]: If you have details about when you've seen that, it would be helpful: I don't think that should be happening (at least not within a single start_date).

[the remark this answers is not in the document]
Comment (rymarczy) [reply python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py:374]: Out flat file test data is a random sample from May 8th of this year. Two trip ids exhibit this behavior (ADDED-1581518542, ADDED-1581518549). 

CSV file with the GTFS-RT Vehicle Position data is attached. 

[non_unique_trip_id.csv](https://github.com/mbta/lamp/files/12432335/non_unique_trip_id.csv)

[the remark this answers is not in the document]
Comment (paulswartz) [reply python_src/src/lamp_py/migrations/versions/performance_manager_prod/001_5d9a7ee21ae5_initial_prod_schema.py:374]: Ah, something like this makes sense for ADDED trips (although probably that reflects a bug in RTR). I'm still not sure why `route_id` needs to be included in the unique constraint though (given that it's possible for some of the other items in there to be non-unique as well), but I'll defer to the Lamp team.

File: python_src/src/lamp_py/performance_manager/l1_cte_statements.py [hunk 3]
```diff
@@ -237,11 +237,7 @@ def trips_for_metrics_subquery(
             sa.func.lead(rt_trips_sub.c.vp_move_timestamp)
             .over(
                 partition_by=rt_trips_sub.c.vehicle_id,
-                order_by=sa.func.coalesce(
-                    rt_trips_sub.c.vp_move_timestamp,
-                    rt_trips_sub.c.vp_stop_timestamp,
-                    rt_trips_sub.c.tu_stop_timestamp,
-                ),
+                order_by=rt_trips_sub.c.vp_move_timestamp,
```
Comment (mzappitello) [inline python_src/src/lamp_py/performance_manager/l1_cte_statements.py:240]: was this change necessary as well if we're running the VACUUM?

Comment (rymarczy) [reply python_src/src/lamp_py/performance_manager/l1_cte_statements.py:240]: This change wasn't necessary, but looking at the field that's being calculated, and how it's used. I don't believe the coalesce is needed. 

I think this business logic was a little bit of a hold over from before Ops Analytics decided they wanted all headways as departure to departure calculations.

**Comment by mzappitello** [comment]
LGTM 🍰
